src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "wallet/coinselection.h"
+#include "util.h"
+#include "utilmoneystr.h"
+
+// Descending order comparator
+struct {
+    bool operator()(const CInputCoin& a, const CInputCoin& b) const
+    {
+        return a.txout.nValue > b.txout.nValue;
+    }
+} descending;
+
+/*
+ * This is the Branch and Bound Coin Selection algorithm designed by Mark Erhardt. It is an exact match algorithm where the exact match
+ * is a range with the lower bound being the value we want to spend and the upper bound being that value plus the additional cost
+ * required to create and spend a change output. To do this, the algorithm builds a binary tree where each node is a UTXO and whether
+ * that UTXO is included or not in the current coin selection. The tree is searched in include-first order. Each UTXO is first included
+ * and the current selection is evaluated for whether it is within the target range. If it is over the threshold, we try excluding
+ * the UTXO previously added and the branch of the tree involving its inclusion is then not explored. This process is repeated until
+ * a solution is found (i.e. selected value falls within the range), the exhaustion limit is reached, or the tree is exhausted and no
+ * solution was found.
+ *
+ * To find the best possible solution, we use a waste metric. The waste metric is defined as the cost to spend the current inputs
+ * now minus the cost to spend the current inputs later, plus the amount exceeding the target value. We search the tree to find the
+ * set of UTXOs which falls within our range and minimizes waste.
+ *
+ * An additional optimization of this algorithm implemented here is a lookahead value which maintains the total value of the UTXO set
+ * of all unexplored UTXOs (i.e. UTXOs that have not yet been included or excluded). This allows us to cut a branch if the remaining
+ * amount is not sufficient to reach our target.
+ *
+ * SelectCoinsBnB Arguments:
+ * const std::vector<CInputCoin>& utxo_pool -> The set of UTXOs that we are choosing from. These UTXOs will be sorted in descending order
+ *                                             by value and the CInputCoins' values are their effective values.
```

Comment:
just say effective value again to clear up ambiguity.

Reply:
Fixed